accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -569,16 +607,241 @@ type `T`, there is an additional type `T*` which is the legacy version of the
+
+The `LEGACY_TOP_MERGE` of more than two types is defined by taking the
+`LEGACY_TOP_MERGE` of the first two, and then recursively taking the
+`LEGACY_TOP_MERGE` of the rest.
+
+If a class which is defined in a legacy library inherits a member with the same
+name from multiple super-interfaces, then error checking is done as usual using
+the legacy typing rules which ignore nullability.  This means that it is valid
+for a legacy class to inherit the same member signature with contradictory
+nullability information. For the purposes of member lookup within a legacy
+library, nullability information is ignored, and so it is valid to simply erase
+the nullability information within the legacy library. When referenced from an
+opted-in library, if there are multiple signatures `T0, ... Tn` for a member
+`m`, and there is a single `Ti` such that `LEGACY_SUBTYPE(Ti, Tk)` for each `k`
+in `0, ..., n`, then the signature of `m` is taken as `Ti`.  Otherwise, the
+signature of `m` for the purposes of member lookup is the `LEGACY_TOP_MERGE` of
+`S0, ..., Sn`, where `Si` is **NORM(`Ti`)**.
+
+We use legacy subtyping when checking inherited member signature coherence in
+classes because opted out libraries may bring together otherwise incompatible
+member signatures without causing an error.
+
+```dart
+// opted_in.dart
+class A {
+  int? foo(int? x) {}
+}
+class B {
+  int foo(int x) {}
+}
+opted_out.dart
+```
+```dart
+// opted out
+// @dart = 2.6
+import 'opted_in.dart';
+
+class C extends A implements B {}
+```
+
+The class `C` is accepted, since the versions of `foo` inherited from `A` and
+`B` are compatible.
+
+If the class `C` is now used within an opted-in library, we must decide what
+signature to ascribe to `foo`.  The `LEGACY_TOP_MERGE` function computes a
+signature for `foo` which preserves as much of the nullability information as
+possible from the inherited members while still arriving at a single coherent
+signature.  In this case, `LEGACY_TOP_MERGE(int? Function(int?), int
+Function(int))` is `int* Function(int*)` and so the following code is accepted:
+
+```dart
+//opted in
+import 'opted_out.dart';
+void test() {
+  new C().foo(null).isEven;
+}
+```
+
+The `LEGACY_TOP_MERGE` also accounts for differences that arise between
+signatures that are otherwise mutual subtypes but which differ in their
+particular choice of top types (e.g. using `dynamic` vs `Object`).  These
+discrepancies are currently resolved simply by choosing the first signature in a
+canonical order.
+
+
+#### Classes defined in opted-in libraries
+
+The `NNBD_TOP_MERGE` of two types `T` and `S` is the unique type `R` defined
+as:
+ - `NNBD_TOP_MERGE(Object?, Object?)  = Object?`
+ - `NNBD_TOP_MERGE(dynamic, dynamic)  = dynamic`
+ - `NNBD_TOP_MERGE(void, void)  = void`
+ - `NNBD_TOP_MERGE(Object?, void)  = void`
+   - And the reverse
+ - `NNBD_TOP_MERGE(dynamic, void)  = void`
+   - And the reverse
+ - `NNBD_TOP_MERGE(Object?, dynamic)  = Object?`
+   - And the reverse
```

Comment:
We have otherwise used the ordering `Object?`, `dynamic`, `void` (not `dynamic`, `Object?`, `void`), for instance in the computation of least upper bounds / greatest lower bounds. Is this intended to change the approach, in order to get a more strict typing? How about the contravariant case where something used to get type `Object*`/`Object?` and with this change it would be `dynamic`? In that case I guess there could be a bunch of method bodies which would suddenly be typed dynamically, because a parameter type is now `dynamic`.

Reply:
> Is this intended to change the approach, in order to get a more strict typing?

It is intended to use a different ordering than we use in upper bound computations, yes.  I think this was our broad preference (to prefer the stricter Object? over dynamic), so if we're going to make a change here, why not make it in the direction we prefer?  We could consider changing UP as well.



> How about the contravariant case where something used to get type `Object*`/`Object?` and with this change it would be `dynamic`

I don't understand this - why does this change cause anything to be `dynamic`?  I believe this change strictly reduces the places where an inherited member signature will have `dynamic`.